### PR TITLE
[QL] Measure API Request Latency for PayPal Flows

### PR DESCRIPTION
**Pull request by jaxdesmarais**
### Summary of changes

- Send `start_time`, `end_time`, and `endpoint` to FPTI for tracking API request latency in the PayPal flow
- Add new `BTCoreAnalytics` enum
- Add new `BTAPITimingDelegate` for sending start/end timing from `BTHTTP` to `BTAPIClient`
- Update `UnitTests`

### Checklist

- [x] Added a changelog entry

- @scannillo @jaxdesmarais

### CODE REVIEW

File: Sources/BraintreeCore/BTHTTP.swift
Comment (scannillo) [reply Sources/BraintreeCore/BTHTTP.swift:4]: Another nit, from [Swift style guide](https://www.swift.org/documentation/api-design-guidelines/):
> Protocols that describe what something is should read as nouns (e.g. Collection).

> Protocols that describe a capability should be named using the suffixes able, ible, or ing (e.g. Equatable, ProgressReporting).

Do we want to keep `Delegate` in the name? I know Apple still does this. I think `HTTPTiming` or `HTTPNetworkTiming` might be more clear?

File: Sources/BraintreeCore/Date+MillisecondTimestamp.swift
Comment (scannillo) [inline Sources/BraintreeCore/Date+MillisecondTimestamp.swift:5]: We use this in a few places so I moved it into an extension. If you hate it, feel free to revert lol

File: UnitTests/BraintreeTestShared/MockAPIClient.swift
Comment (scannillo) [inline UnitTests/BraintreeTestShared/MockAPIClient.swift:99]: Because of this, our assertions added in `BTPayPalClient_Tests` are only testing this Mock behavior, and not executing any of our new code added in BTAPIClient & BTHTTP.

We _could_ override `fetchAPITiming()` in this Mock class instead, but it'll be a similar situation of only passing based on added Mock code.

I would update BTAPIClient & BHTTP tests instead, and then the best we can do at the PayPalClient level is assert that `shouldSendAPIRequestLatency` is true.

File: Sources/BraintreeCore/BTHTTP.swift
Comment (jaxdesmarais) [reply Sources/BraintreeCore/BTHTTP.swift]: Good call, updated here: https://github.com/braintree/braintree_ios/pull/1292/commits/116d09590f8fda77152b0deab04cd35f554c77e5

File: Sources/BraintreeCore/BTHTTP.swift
Comment (jaxdesmarais) [reply Sources/BraintreeCore/BTHTTP.swift]: For sure, updated here: https://github.com/braintree/braintree_ios/pull/1292/commits/550c7226c10b42e62406f0d36f9953098ab4e372

File: Sources/BraintreeCore/Date+MillisecondTimestamp.swift
Comment (agedd) [inline Sources/BraintreeCore/Date+MillisecondTimestamp.swift:5]: small nit: 
```suggestion
    var utcTimestampMilliseconds: Int {
```

File: Sources/BraintreePayPal/BTPayPalClient.swift
Comment (agedd) [inline Sources/BraintreePayPal/BTPayPalClient.swift:48]: qq: do we always want this to be set to true?

Comment (agedd) [inline Sources/BraintreePayPal/BTPayPalClient.swift:48]: qq: do we always want this to be set to true? 👀

File: Sources/BraintreeCore/Date+MillisecondTimestamp.swift
Comment (jaxdesmarais) [reply Sources/BraintreeCore/Date+MillisecondTimestamp.swift]: I had noticed the same, updated here: https://github.com/braintree/braintree_ios/pull/1292/commits/e2aa0ea330939b6cb7ffec336c88fea90167854c

File: Sources/BraintreePayPal/BTPayPalClient.swift
Comment (jaxdesmarais) [reply Sources/BraintreePayPal/BTPayPalClient.swift:48]: For PayPal currently yes we want to send this for all `BTPayPalClient` flows so we can track this across both current and future QL initiatives. This branch will get merged into our other feature branches to ensure we are tracking these events there as well.

File: UnitTests/BraintreeTestShared/MockAPIClient.swift
Comment (jaxdesmarais) [reply UnitTests/BraintreeTestShared/MockAPIClient.swift]: I updated `BTPayPalClient_Tests` to just test setting `shouldSendAPIRequestLatency` here: https://github.com/braintree/braintree_ios/pull/1292/commits/8af3d0b4452f740f97b6feb73d4fa9ad549defba. For `BTAPIClient_Tests` we are going to run into the same issue, currently those tests when testing analytics ([example here](https://github.com/braintree/braintree_ios/blob/2120f0b705d4fd1f6173b431d71a7d9705a7d096/UnitTests/BraintreeCoreTests/BTAPIClient_Tests.swift#L406-L414)) use `FakeAnalyticsService` ([here](https://github.com/braintree/braintree_ios/blob/main/UnitTests/BraintreeCoreTests/Analytics/FakeAnalyticsService.swift)) which just sends whatever event you tell it to send. So we would end up just testing our mock there as well.

**Comment by scannillo** [comment]
Based on[ this diagram ](https://developer.apple.com/documentation/foundation/urlsessiontasktransactionmetrics#3162615)from Apple’s docs, maybe requestStart isn’t what we want, but rather fetchStart? The DNS lookup, TCP & TLS setup aren’t included in requestStart

**Comment by jaxdesmarais** [comment]
> Based on[ this diagram ](https://developer.apple.com/documentation/foundation/urlsessiontasktransactionmetrics#3162615) from Apple’s docs, maybe requestStart isn’t what we want, but rather fetchStart?
> 
> The DNS lookup, TCP & TLS setup aren’t included in requestStart.

Nice find, updated here: https://github.com/braintree/braintree_ios/pull/1292/commits/7ffd8ab6c1a473264a6e4ce9d090dc859b260c99

File: UnitTests/BraintreeTestShared/MockAPIClient.swift
Comment (scannillo) [reply UnitTests/BraintreeTestShared/MockAPIClient.swift]: But we can assert that `FakeAnalyticsService.sendAnalyticsEvent()` was called with the expected values, to test the logic that lives within `BTAPIClient.fetchAPITiming()`.

Rough outline like:
* when networkTimingDelegate.fetchAPITiming() delegate method invoked, sends analytic event
* when delegate method invoked, and analytics path, does not send analytic event
* when delegate method invoked, and shouldSendAPILatency = false, does not send analytic event
* when delegate method invoked, and path contains merchantID, strips merchantID from posted JSON body tag param

The easiest way would be to call `sut.fetchAPITiming()` directly in each test. The probably "right" way would be to set a MockTimingDelegate on sut and trigger the delegate call that way via the mock. Not sure it's worth the setup for the later option at this point though.

Comment (jaxdesmarais) [reply UnitTests/BraintreeTestShared/MockAPIClient.swift]: > But we can assert that FakeAnalyticsService.sendAnalyticsEvent() was called with the expected values, to test the logic that lives within BTAPIClient.fetchAPITiming().

This method does not behave at least how I would expect, because my initial thought was the same. If we take a look at the test `testSendAnalyticsEvent_whenCalled_callsAnalyticsService` ([here](https://github.com/braintree/braintree_ios/blob/2120f0b705d4fd1f6173b431d71a7d9705a7d096/UnitTests/BraintreeCoreTests/BTAPIClient_Tests.swift#L406-L414)) it is calling `FakeAnalyticsService.sendAnalyticsEvent()` with `blahblah` and asserting that the `mockAnalyticsService.lastEvent` equals `blahblah`. 

Which will always be true because that's the input we are giving it. While it is an override we are not actually hitting the override unless we call it directly so our call to `BTAPIClient.fetchAPITiming()` never actually hits `FakeAnalyticsService.sendAnalyticsEvent()`. We can take a look at this a bit later if that doesn't make sense.

Comment (jaxdesmarais) [reply UnitTests/BraintreeTestShared/MockAPIClient.swift]: Paired on this because I was overthinking it, updated here: https://github.com/braintree/braintree_ios/pull/1292/commits/ff199133838acd50aa135b63a273fa2e31103b2c

File: UnitTests/BraintreeTestShared/MockAPIClient.swift
Comment (scannillo) [inline UnitTests/BraintreeTestShared/MockAPIClient.swift:16]: We can remove this

File: UnitTests/BraintreePayPalTests/BTPayPalClient_Tests.swift
Comment (scannillo) [inline UnitTests/BraintreePayPalTests/BTPayPalClient_Tests.swift:21]: We can remove this

**Review by scannillo — APPROVED** [review]
Changes look good! But would mark as "DO NOT MERGE" until #1295 is merged.

File: UnitTests/BraintreeTestShared/MockAPIClient.swift
Comment (jaxdesmarais) [reply UnitTests/BraintreeTestShared/MockAPIClient.swift]: Done: https://github.com/braintree/braintree_ios/pull/1292/commits/58768861720cbaa7e3c1b16aba7fb6102b82b0c7

**Comment by scannillo** [comment]
One thing came to my mind - once the batching logic is in place, maybe we should just send this analytic event for all API requests (not just PP)? Since our batching is time based and not based on the # of events, I don't think it will have too much impact on consuming too much networking bandwidth.

Do we think those metrics would also be useful to have?

File: Sources/BraintreeCore/BTAPIClient.swift
Comment (richherrera) [inline Sources/BraintreeCore/BTAPIClient.swift:480]:
```suggestion
            if let clientToken, let baseURL {
                http = BTHTTP(url: baseURL, authorizationFingerprint: clientToken.authorizationFingerprint)
            } else if let tokenizationKey, let baseURL {
                http = BTHTTP(url: baseURL, tokenizationKey: tokenizationKey)
            }
            
            http?.networkTimingDelegate = self
```

**Review by richherrera — APPROVED** [review]

**Review by agedd — APPROVED** [review]
lgtm! 🚀

**Comment by scannillo** [comment]
@jaxdesmarais following up here - we talked IRL about merging this in anyways, even if batching isn't complete. Do we want to run a few comparison tests using OS Signpost to make sure this change isn't adding too much additional time?

Getting actual data for these analytics from real merchant usage will help us a ton

**Comment by jaxdesmarais** [comment]
> @jaxdesmarais following up here - we talked IRL about merging this in anyways, even if batching isn't complete. Do we want to run a few comparison tests using OS Signpost to make sure this change isn't adding too much additional time?
> 
> Getting actual data for these analytics from real merchant usage will help us a ton

For sure, I am using branch `api-request-latency-signpost` for testing. Results are on cellular:
- Production: 1.14s, 628.27ms, 957.79ms, 495.53ms, 395.75ms, 864.03ms, 486.59ms, 482.74ms, 717.69ms, 742.30ms = 678.96 average
 
Results on `main`:
- Production: 1.02s, 962.32ms, 571.74ms, 547.22ms, 443.80ms, 610.82ms, 461.21ms, 493.29ms, 432.27ms, 409.00ms = 593.37 average

So it looks like on average this would add 85.59ms of latency without batching. Curious if your results are the same.

**Comment by scannillo** [comment]
Here are my results. I ran them on WiFi for more consistency b/w trials:

Screenshot 2024-05-14 at 1 24 27 PM

**Comment by jaxdesmarais** [comment]
Are we okay with that variance on stable vs slightly less stable connections to merge this in/release before batching? I agree the sooner we can get more data on these hops the better.

**Comment by jaxdesmarais** [comment]
For reference here are my local tests with merging this branch into `batch-analytics` locally:

- Production wifi: 572.52ms, 319.12ms, 426.20ms, 399.95ms, 320.96ms, 404.85ms, 614.79ms, 439.32ms, 439.06ms, 522.78ms = 449.95 average
- Production cellular: 931.14ms, 540.38ms, 548.63ms, 712.02ms, 769.38ms, 545.93ms, 542.22ms, 654.36ms, 692.68ms, 600.24ms = 653.63 average

**Comment by jaxdesmarais** [comment]
Discussed with @scannillo irl and removing the "do not merge" tag as even in the worst case our latency is minimally impacted. This will allow us to get additional merchant data on these hops and we will fast follow with batching.

**Pull request merged**
